Manager/src/main/java/io/raspberrywallet/manager/bitcoin/Bitcoin.java

```diff
@@ -51,58 +80,119 @@ private NetworkParameters parseNetworkFrom(Configuration.BitcoinConfig bitcoinCo
-                .setBlockingStartup(false)
-                .setUserAgent("RaspberryWallet", "1.0");
-
-        if (seed != null)
-            kit.restoreWalletFromSeed(seed);
+        if (peerGroup != null && peerGroup.isRunning()) {
+            // Shut down synchronization before setting up new wallet
+            ListenableFuture future = peerGroup.stopAsync();
+            future.addListener(setupWalletFromBackup, Executors.newSingleThreadExecutor());
+        } else
+            setupWalletFromBackup.run();
     }
 
-    public WalletAppKit getKit() throws WalletNotInitialized {
-        if (kit == null) throw new WalletNotInitialized();
-        return kit;
+
+    /**
+     * Download blockchain from PeerGroup discovered by DnsDiscovery on the NetworkParams
+     *
+     * @param wallet wallet to index transactions for
+     */
+    private void synchronizeWalletBlocking(Wallet wallet) throws BlockStoreException, IOException {
+        long start = System.currentTimeMillis();
+
+        InputStream checkpoints = CheckpointManager.openStream(params);
+        CheckpointManager.checkpoint(params, checkpoints, blockStore, wallet.getEarliestKeyCreationTime());
+
+        BlockChain chain = new BlockChain(params, wallet, blockStore);
+
+        peerGroup = new PeerGroup(params, chain);
+        peerGroup.setUserAgent("RaspberryWallet", "1.0");
+        peerGroup.addPeerDiscovery(new DnsDiscovery(params));
+        peerGroup.addAddress(new PeerAddress(params, InetAddress.getLocalHost()));
+        peerGroup.start();
+        peerGroup.downloadBlockChain();
+
+        long total = System.currentTimeMillis() - start;
+        Logger.info(String.format("Synchronization took %.2f secs", (double) total / 1000.0));
     }
 
+    @NotNull
     public Wallet getWallet() throws WalletNotInitialized {
-        if (kit == null) throw new WalletNotInitialized();
-        return kit.wallet();
+        Wallet _wallet = wallet;
+        if (_wallet == null) throw new WalletNotInitialized();
+        return _wallet;
     }
 
-    void importKey(byte[] keyBytes) throws WalletNotInitialized {
-        importKey(ECKey.fromPrivate(keyBytes));
+    public void ensureWalletInitialized() throws WalletNotInitialized {
+        if (wallet == null) throw new WalletNotInitialized();
     }
 
-    private void importKey(ECKey key) throws WalletNotInitialized {
-        getWallet().importKey(key);
+    /**
+     * This method encrypt wallet, saves it onto disk and decrypt it back so it become again usable
+     *
+     * @param key key used to encrypt wallet before saving onto disk
+     * @throws IOException          when the problem with saving wallet occurs
+     * @throws WalletNotInitialized when you try to save not initialized wallet
+     */
+    public void saveEncryptedWallet(KeyParameter key) throws IOException, WalletNotInitialized {
+        ensureWalletInitialized();
+
+        encryptWallet(key);
+        getWallet().saveToFile(walletFile);
```

Comment:
Since `ensureWalletInitialized();` checks if wallet is initialized, then there is no need to do it again: 
```
wallet.saveToFile(walletFile);
```
But I'm not sure, if it is not made like that, due to BitcoinJ behaviour, that may change wallet state between calls

Reply:
Wallet Initialized == wallet loaded into memory and operable.

saveToFile() persist the wallet from memory to disk.

Reply:
okay, now I understand